alternator/conditions.cc

```diff
@@ -143,9 +143,50 @@ static void verify_operand_count(const rjson::value* array, const size_check& ex
     }
 }
 
+struct rjson_engaged_ptr_comp {
+    bool operator()(const rjson::value* p1, const rjson::value* p2) const {
+        return rjson::single_value_comp()(*p1, *p2);
+    }
+};
+
+// It's not enough to compare underlying JSON objects when comparing sets,
+// as internally they're stored in an array, and the order of elements is
+// not important in set equality. See issue #5021
+static bool check_EQ_for_sets(const rjson::value& s1, const rjson::value& s2) {
+    const rjson::value* set1 = unwrap_set(s1).second;
+    const rjson::value* set2 = unwrap_set(s2).second;
+
+    if (!set1 || !set2) {
```

Comment:
Please add a comment what you're checking here and why, it's very unclear. What does !set1 even mean? You only call this function if both objects are really sets, and of the same type. Can !set1 happen - when? And why is !set && !set2 the proper return value?

Reply:
@psarna I'm still curious about the answer to this question.

Reply:
Ah, sorry, missed the question yesterday. I just didn't assume here that the caller already checked if both of the items are sets. In case a non-set is passed to unwrap_set(), a nullptr will be returned, and the semantics I had in mind is that if any of the items is not a set, it will return true only if both are non-sets. As I'm writing this response though, I see how flawed this reasoning is, because the fact that both items are not sets does not by any means make them equal. I'll just fix this by throwing an internal error if (!set! || !set2) - as a light version of asserting that both items passed here should really be sets.

Reply:
Sure. Let me know when you push a new version, and I can commit it to next.